File: frontend/src/api/imports.ts
import { api, download } from "./client";

/**
 * Loading a spreadsheet somebody exported from something else (§29).
 *
 * Five things worth knowing before reading the types.
 *
 * **The steps are the model's, not the page's.** `step` is stored on the run,
 * so a draft reopens where it was left rather than wherever the data implies —
 * a run whose mapping happens to be complete but which nobody has looked at
 * should land on the preview, not skip past it.
 *
 * **`column_mapping` is `{column: field}`,** which reads backwards until you
 * notice why: the file is the thing with duplicates in it. Two columns can
 * want the same field and the server refuses that, and keying by column is
 * what makes the refusal expressible.
 *
 * **The counts add up.** `total_rows = valid + invalid + skipped` once the
 * rows have been validated, and `skipped` means one thing: after mapping, the
 * row has nothing in any mapped column. Before validation they are all zero,
 * because they are not yet facts about anything.
 *
 * **`line` is the line in the file**, not an index. Row 1 is the header, so
 * the first data row is 2 — which is the number somebody needs to find it in
 * the spreadsheet they are about to fix.
 *
 * **Nothing is written until the preview has been seen**, and then it is
 * all-or-nothing: a failure part-way leaves no records at all. `can_execute`
 * comes from the server for the same reason every other control does.
 *
 * **Letting one go takes two presses**, the rule `/exports` and `/mail` use:
 * the file first, the record second, with `removed` saying which happened.
 *
 * See `services/imports.py`.
 */

export type ImportStatus =
  | "DRAFT"
  | "VALIDATED"
  | "RUNNING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

export type ImportStep = "UPLOAD" | "MAPPING" | "PREVIEW" | "EXECUTE" | "DONE";

/** One column as found in the file, with samples so two can be told apart. */
export interface DetectedColumn {
  index: number;
  name: string;
  samples: string[];
}

/** One thing wrong with one row, in the four terms needed to fix it. */
export interface RowProblem {
  /** The line in the file. The header is line 1. */
  line: number;
  column: string;
  field?: string;
  /** The value as it was written, so the report reads beside the sheet. */
  value: string;
  message: string;
}

/** One staged row, as the file has it and as the mapping reads it. */
export interface PreviewRow {
  line: number;
  source: Record<string, string>;
  values: Record<string, string | undefined>;
  problems: RowProblem[];
}

export interface ImportRun {
  id: string;
  reference: string;
  filename: string;
  target_entity: string;
  target_label: string;
  status: ImportStatus;
  step: ImportStep;
  delimiter: string;
  total_rows: number;
  valid_rows: number;
  invalid_rows: number;
  skipped_rows: number;
  imported_rows: number;
  detected_columns: DetectedColumn[];
  /** `{column: field}` — see the module note on why that way round. */
  column_mapping: Record<string, string>;
  /** Required fields with no column yet. Empty means ready to validate. */
  unmapped_required: string[];
  error_count: number;
  created_at: string | null;
  completed_at: string | null;
  can_execute: boolean;
  /** Whether letting it go would do anything — false only while it is writing. */
  can_discard: boolean;
  /**
   * Whether it still holds the staged file, and so which of `discard`'s two
   * presses the next one is. Said by the server rather than inferred: the
   * `/exports` page inferred the equivalent from `size_bytes` and could never
   * reach its own second press.
   */
  holds_file: boolean;
}

export interface ImportDetail extends ImportRun {
  errors: RowProblem[];
  preview: PreviewRow[];
  preview_total: number;
  /** Present on the answer to `begin`: what the file turned out to be. */
  dialect?: {
    delimiter: string;
    label: string;
    consistent: boolean;
    /** A sentence shown to the reader, so a wrong guess is visible. */
    note: string;
  };
  blank_rows?: number;
  ran_as?: string;
}

/** One field a column may be mapped onto, and what it accepts. */
export interface ImportField {
  name: string;
  label: string;
  kind: string;
  required: boolean;
  choices: string[];
  minimum: number | null;
  maximum: number | null;
  /** Which dataset a foreign key points at, so the wizard can say so. */
  references: string;
}

export interface ImportTarget {
  key: string;
  label: string;
  description: string;
  fields: ImportField[];
  required: string[];
}

export interface ImportCatalogue {
  targets: ImportTarget[];
  delimiters: Array<{ key: string; label: string }>;
  max_rows: number;
  max_bytes: number;
  preview_rows: number;
  open_limit: number;
  open: number;
  statuses: Array<{ key: ImportStatus; count: number }>;
  steps: ImportStep[];
  total: number;
}

export interface ImportsPage {
  items: ImportRun[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  facets: Record<string, Array<{ value: string; count: number }>>;
  columns: string[];
}

export interface ImportQuery {
  [key: string]: unknown;
  q?: string;
  status?: string;
  page?: number;
  page_size?: number;
}

export const importsApi = {
  catalogue: (signal?: AbortSignal) =>
    api.get<ImportCatalogue>("/imports/catalogue", { signal }),
  list: (params: ImportQuery = {}, signal?: AbortSignal) =>
    api.get<ImportsPage>("/imports", { params, signal }),
  entry: (id: string, signal?: AbortSignal) =>
    api.get<ImportDetail>(`/imports/${id}`, { signal }),
  /**
   * Send the file's text. Not a presigned upload: the API has to parse it, so
   * routing the bytes through object storage and back would move them through
   * the same worker twice.
   */
  begin: (input: {
    target_entity: string;
    filename: string;
    content: string;
    delimiter?: string;
  }) => api.post<ImportDetail>("/imports", input),
  /** Set the mapping and validate every row — one call, because it is one thought. */
  map: (id: string, columnMapping: Record<string, string>) =>
    api.put<ImportDetail>(`/imports/${id}/mapping`, { column_mapping: columnMapping }),
  /** Write the valid rows. All of them or none of them. */
  execute: (id: string) => api.post<ImportDetail>(`/imports/${id}/execute`, {}),
  /**
   * Let one go. The first press abandons the draft and drops the staged file;
   * a second removes the record. `removed` says which happened.
   */
  discard: (id: string) =>
    api.delete<ImportRun & { removed?: boolean }>(`/imports/${id}`),
  /** The error report, saved as a file to fix in the spreadsheet. */
  problems: (id: string, reference: string) =>
    download(`/imports/${id}/problems`, { fallbackName: `${reference}-problems.csv` }),
};

File: backend/src/api/imports.py
"""Import endpoints (§29).

One permission, `records.import` — narrower than `records.create` on purpose:
creating one record is a form somebody filled in and can see, while importing
five thousand is an act nobody reads row by row. It is also what `/import`
declares in the navigation, and the route guard reads the navigation entry, so
the two must agree.

The operations are the wizard's steps, and the shape of the set is the point:
**catalogue** says what may be imported into and what each field accepts,
**begin** takes the file and reads it, **mapping** says which column is which
field *and validates every row in the same call*, **item** is the preview,
**execute** writes them, **problems** hands back the error report as a file,
and **item**'s `DELETE` lets one go — the file on the first press, the
record on the second.

Two things deliberately absent. There is no "validate" call: changing a mapping
changes which rows are wrong, so making somebody press Validate afterwards
would be a step whose answer is already known. And there is no partial execute
— see the service docstring on why half an import is the outcome this flow
exists to prevent.
"""

from __future__ import annotations

from typing import Any

from src.core.auth import json_body, me, requires
from src.core.db import session_scope
from src.services import imports as service


@requires("records.import")
def catalogue(app=None, operation: str = "", request=None, **_: Any):
    """What can be imported into, what each field accepts, and the limits.

    The fields come from the same `Writable` declarations an edit form is
    rendered from, so a field that becomes writable becomes importable the same
    day — and the wizard cannot offer one the create would refuse.
    """
    with session_scope() as session:
        return service.catalogue(session, principal=me()), 200


@requires("records.import")
def collection(app=None, operation: str = "", request=None, **_: Any):
    """My imports, newest first, filtered and faceted in PostgreSQL."""
    args = request.args if request is not None else {}
    with session_scope() as session:
        return service.listing(session, args, principal=me()), 200


@requires("records.import")
def begin(app=None, operation: str = "", request=None, **_: Any):
    """Accept a file, read it, and propose a mapping. The first step.

    The CSV arrives as text in the body rather than through a presigned upload,
    because the API has to parse it: sending it to object storage and fetching
    it back would move the same bytes through the same worker twice. Bounded by
    `core/importer.MAX_BYTES` and refused before anything is stored.
    """
    with session_scope() as session:
        return service.begin(session, json_body(), principal=me()), 201


@requires("records.import")
def item(app=None, operation: str = "", request=None, import_id: str = "", **kwargs: Any):
    """One import with a page of its rows and their problems — the preview.

    `DELETE` takes two presses, as `/exports` and `/mail` do: the first
    abandons the draft and drops the staged copy of somebody's spreadsheet,
    the second removes the record. `removed` in the answer says which
    happened, and the audit trail keeps the history either way.
    """
    identifier = import_id or str(kwargs.get("import_id") or "")
    method = (request.method if request is not None else "GET").upper()
    with session_scope() as session:
        if method == "DELETE":
            return service.discard(session, identifier, principal=me()), 200
        return service.entry(session, identifier, principal=me()), 200


@requires("records.import")
def mapping(app=None, operation: str = "", request=None, import_id: str = "", **kwargs: Any):
    """Set which column is which field, and validate every row.

    One call for both, because they are one thought: a mapping change changes
    which rows are wrong. Answers with the preview, so the page has the result
    of the decision it just made.
    """
    identifier = import_id or str(kwargs.get("import_id") or "")
    with session_scope() as session:
        return service.map_columns(session, identifier, json_body(), principal=me()), 200


@requires("records.import")
def execute(app=None, operation: str = "", request=None, import_id: str = "", **kwargs: Any):
    """Write the valid rows — all of them or none of them.

    Runs off the request, as a queued export does (§30), and answers with the
    run marked RUNNING plus the mechanism producing it. A small import is
    finished before the page redraws; a large one is followed.
    """
    identifier = import_id or str(kwargs.get("import_id") or "")
    with session_scope() as session:
        return service.execute(session, identifier, principal=me()), 202


@requires("records.import")
def problems(app=None, operation: str = "", request=None, import_id: str = "", **kwargs: Any):
    """The error report, as a CSV.

    A file rather than a list, because the rows have to go back to the
    spreadsheet they came from: line, column, the value as written, and the
    reason.
    """
    identifier = import_id or str(kwargs.get("import_id") or "")
    with session_scope() as session:
        return service.problems(session, identifier, principal=me())
